### PR TITLE
[Docs] upgrade/chain halt recovery

### PR DESCRIPTION
**Pull request by okdas**
## Summary

Performed the first upgrade on the Alpha TestNet. Add some documentation changes to prevent some issues in the future.

## Issue

N/A

## Type of change

Select one or more from the following:

- [ ] New feature, functionality or library
- [ ] Consensus breaking; add the `consensus-breaking` label if so. See #791 for details
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [x] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

**Comment by okdas** [comment]
Items to add:
- Configure cosmovisor to create backups before upgrade
- Have more bump validators to 5.
- Instructions on how to test the upgrade locally. Thinking a script w/o LocalNet.

**Comment by Olshansk** [comment]
@okdas Let's add the details here as well: https://x.com/olshansky/status/1846211059989778741

**Comment by okdas** [comment]
@Olshansk can I get a review on this PR, please?

I think I want to investigate/add information about the new Cosmovisor feature mentioned in `cosmos-sdk` slack, but I'll look into that later. :)

File: docusaurus/docs/protocol/upgrades/contigency_plans.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+:::
+
+### Option 2: The migration is stuck
+
+If the migration is stuck, there's always a chance the state has been mutated for
+the upgrade but the migration didn't complete.
+
+In such a case, we need to:
+
+- Roll back validators to the backup (a snapshot is taken by `cosmovisor` automatically prior to upgrade, if `UNSAFE_SKIP_BACKUP` is set to `false`).
```
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/contigency_plans.md:58]: Please link to the appropriate doc/section for this

File: docusaurus/docs/protocol/upgrades/contigency_plans.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+- Skip the upgrade handler and store migrations with `--unsafe-skip-upgrade=$upgradeHeightNumber`.
```
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/contigency_plans.md:59]: Please show full command (or link to script). 

I don't know what this means

File: docusaurus/docs/protocol/upgrades/contigency_plans.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+- Document and add `--unsafe-skip-upgrade=$upgradeHeightNumber` to the scripts so the next time somebody tries to sync the network from genesis
```
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/contigency_plans.md:60]: Which script? #PUC

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md [hunk 4]
```diff
@@ -91,18 +141,89 @@ Note that for local testing, `cosmovisor` won't pull the binary from the info fi
...
+8. Verify the plan is active
 
-2. Check and update the `tools/scripts/upgrades/cosmovisor-start-node.sh` to point to the correct binaries:
+   ```bash
+   poktrolld query upgrade plan
+   ```
 
-   - The old binary should be compiled to work before the upgrade.
-   - The new binary should contain the upgrade logic to be executed immediately after the node is started using the new binary.
+9. Wait until the height is reached and the old node dies due to the error: `ERR UPGRADE "v0.0.9-2" NEEDED at height`, which is expected.
```
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/upgrade_procedure.md:217]: I did a little bit of partial cleanup, but can you do another look through this and make it so copy-pastable that any idiot can follow?

For example, instead of "from the new version", just add the appropriate `cd ../...` in there.

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md [hunk 5]
```diff
@@ -91,18 +141,89 @@ Note that for local testing, `cosmovisor` won't pull the binary from the info fi
...
+11. In the **NEW VERSION GIT REPO** you can build binaries using `go_develop`, `ignite_release` and `ignite_release_extract_binaries` make targets.
+12. Start the new version from the **NEW VERSION REPO**:
 
-3. Run `bash tools/scripts/upgrades/cosmovisor-start-node.sh` to wipe the `~/.poktroll` directory and place binaries in the correct locations.
+    ```bash
+    ./release_binaries/poktroll_darwin_arm64 start
+    ```
 
-4. Execute the transaction as shown in [Submitting the upgrade on-chain](#submitting-the-upgrade-on-chain) section above.
+13. Observe the behavior. Your node should go through the upgrade process and start using the new version.
```
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/upgrade_procedure.md:226]: How do I observe the behaviour?

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md (hunk 4 above)
Comment (Olshansk) [inline docusaurus/docs/protocol/upgrades/upgrade_procedure.md:217]: How do I observe the height?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+**This is preferred way of resolving the consensus-breaking issues**.
+
+:::
+
+Since the chain is not moving, **it is impossible** to issue an automatic upgrade with an upgrade plan.
+
+Instead, we need **social consensus** to manually replace the binary and get the chain moving.
+
+Currently this involves synching the network from genesis breaking a way to sync the network from genesis without human interaction, but there are some plans to make the process less painful in the future.
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md:50]: LOL the original sentence makes more sense. Let me rephrase..

File: tools/installer/full-node.sh [hunk 7]
```diff
@@ -138,12 +138,15 @@ setup_poktrolld() {
         exit 1
     fi
 
-    # Use the direct download link for the latest release
-    LATEST_RELEASE_URL="https://github.com/pokt-network/poktroll/releases/latest/download/poktroll_linux_${ARCH}.tar.gz"
+    # Get the version genesis started from
+    POKTROLLD_VERSION=$(curl -s https://raw.githubusercontent.com/pokt-network/pocket-network-genesis/master/poktrolld/testnet-validated.init-version)
```
[the remark this answers is not in the document]
Comment (okdas) [reply tools/installer/full-node.sh:143]: I rephrased this part, but I don't fully understand what you're asking.
If you think that there's a better way, please let me know. I thought we could use the `app_version` from `genesis.json` (looks like this: `"app_version": "0.0.9"`) but decided I'd rather store this information in a separate file.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+and use the same version of the software.
+
+## Resolving halts during a network upgrade
+
+If the halt is caused by the network upgrade, it is possible the solution can be as simple as
+skipping an upgrade (i.e. `unsafe-skip-upgrade`) and creating a new (fixed) upgrade.
+
+Read more about [upgrade contingency plans](../../protocol/upgrades/contigency_plans.md).
+
+### Manual binary replacement (preferred)
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md:38]: There are no binaries to point to. Rephrased.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+8. Wait for the network to reach the height of the previous ledger (`104`+)
+9. Allow validators to open their network to full nodes again.
+   - Note that full nodes will need to perform the rollback or use a snapshot as well.
+
+#### Step 5: Data rollback - retrieving snapshot at a specific height
+
+There are two ways to get a snapshot from a prior height:
+
+1. Use `poktrolld rollback --hard` repeately until the command responds with the desired block number.
+2. Use a snapshot and start the node with `--halt-height=100` parameter so it only syncs up to certain height and then gracefully shuts down.
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md]: Added an example, but I don't think it is helpful. There's no one command that will work for everybody.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+   - If using `cosmosvisor`, the node will wait to replace the binary
+8. Wait for the network to reach the height of the previous ledger (`104`+)
+9. Allow validators to open their network to full nodes again.
+   - Note that full nodes will need to perform the rollback or use a snapshot as well.
+
+#### Step 5: Data rollback - retrieving snapshot at a specific height
+
+There are two ways to get a snapshot from a prior height:
+
+1. Use `poktrolld rollback --hard` repeately until the command responds with the desired block number.
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md]: Added, but honestly, I prefer one ` here.

File: docusaurus/docs/protocol/upgrades/contigency_plans.md (hunk 1 above)
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/contigency_plans.md]: Added a link to cosmovisor docs.

File: docusaurus/docs/protocol/upgrades/contigency_plans.md (hunk 2 above)
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/contigency_plans.md]: Just like with `--halt-height=` there's no command that will work for all use cases. I'll add an example, but it's not very usable.

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md (hunk 4 above)
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:217]: How do I know what the new version path on the developer's system? It might be a primary `poktroll` repo on the developer machine, it might not be. ¯\\\_(ツ)_/¯

Comment (Olshansk) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:217]: #PUC? Screenshot? Video?

Comment (Olshansk) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:217]: Maybe an instruction where they run `poktroll version` to ensure it's correct?

It's one of the steps in the upgrade playbook for Morse:

Screenshot 2024-10-24 at 12 18 01 PM

File: docusaurus/docs/protocol/upgrades/contigency_plans.md (hunk 2 above)
Comment (Olshansk) [reply docusaurus/docs/protocol/upgrades/contigency_plans.md]: #PUC

- Bullet points
- Consider making it a separate section below

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md]: We can change it back if need be.

You know how they say "the customer is always right"?

I'm the customer of this document.

**Comment by Olshansk** [comment]
@okdas Friendly reminder on this when you're back.

Making it easier to inspect/find/identify chain halts on testnet would be great as well!

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md (hunk 4 above)
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md]: Started by writing a cheat sheet, ended up changing the whole section.

**Comment by okdas** [comment]
> @okdas Lmk when this is ready for another look.

It is ready now! Thanks!

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md [hunk 11]
```diff
@@ -67,42 +70,143 @@ An upgrade transaction includes a [Plan](https://github.com/cosmos/cosmos-sdk/bl
...
+   ```bash
+   make go_develop ignite_release ignite_release_extract_binaries
+   ```
+   :::note
+   The binary produced by these commands in the old repo should result in the same binary as it was downloaded from [production releases](https://github.com/pokt-network/poktroll/releases), however you might consider using 
+   :::
+
+5. **(`old` repo)** - Clean up and generate an empty genesis using the old version.
+   ```bash
+   rm -rf ~/.poktroll && ./release_binaries/poktroll_darwin_arm64 comet unsafe-reset-all && make localnet_regenesis
```
Comment (bryanchriswhite) [inline docusaurus/docs/protocol/upgrades/upgrade_procedure.md:174]: make target?

**Review by bryanchriswhite — DISMISSED** [review]
Great docs @okdas! :raised_hands: :heart_eyes:

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+### Rollback, fork and upgrade
+
+:::info
+
+These instructions are only relevant to Pocket Network's Shannon release.
+
+We do not currently use `x/gov` and on-chain voting for upgrades.
+
+Instead, our DAO votes on upgrades off-chain and the Foundation executes
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md:93]: DAO = All people who participate in voting.
Foundation = PNF (e.g. Mike), who execute the will of DAO voters on their behalf. Let me update that part.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+transactions on their behalf.
+
+:::
+
+**Performing a rollback is analogous to forking the network at the older height.**
+
+This should be avoided unless absolutely necessary.
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md:100]: Hm, we recommend the `Manual binary replacement (preferred)`, and do not recommend `Rollback, fork and upgrade`. Added more information to help with confusion.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+2. [Create a release](../../protocol/upgrades/release_process.md).
+3. [Prepare an upgrade transaction](../../protocol/upgrades/upgrade_procedure.md#writing-an-upgrade-transaction) to the new version.
+4. Get the Validator set off the network **3 blocks** prior to the height of the chain halt. For example:
+   - Assume an issue at height `103`
+   - Get the validator set at height `100`
+   - Submit an upgrade transaction at `101`
+   - Upgrade the chain at height `102`
+   - Avoid the issue at height `103`
+5. Ensure all validators rolled back to the same height and use the same snapshot - ([how to get the snapshot](#step-5-data-rollback---retrieving-snapshot-at-a-specific-height))
+   - The snapshot should be imported into each Validator's data directory
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md]: Correct

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md (hunk 11 above)
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:189]: With these commands, I want a developer to 100% understand what they are doing, checking all arguments.

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md [hunk 15]
```diff
@@ -67,42 +70,143 @@ An upgrade transaction includes a [Plan](https://github.com/cosmos/cosmos-sdk/bl
...
+
+7. **(`old` repo)** Start the node:
+   ```bash
+   ./release_binaries/poktroll_darwin_arm64 start
+   ```
+   The validator node should run and produce blocks as expected.
+
+8. **(`old` repo)** Submit the upgrade transaction. **Note:** The upgrade height in the transaction should be higher than the current block height. Adjust and submit if necessary:
+   ```bash
+   ./release_binaries/poktroll_darwin_arm64 tx authz exec tools/scripts/upgrades/local_test_v0.2.json --from=pnf
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:205]: With these commands, I want a developer to 100% understand what they are doing, checking all arguments.

File: docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+#### Step 5: Data rollback - retrieving snapshot at a specific height
+
+There are two ways to get a snapshot from a prior height:
+
+1. Execute
+   ```bash
+   poktrolld rollback --hard
+   ```
+   repeately until the command responds with the desired block number.
+2. Use a snapshot and start the node with `--halt-height=100` parameter so it only syncs up to certain height and then
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/develop/developer_guide/recovery_from_chain_halt.md:162]: No, added more information about that. Thanks!

File: docusaurus/docs/protocol/upgrades/upgrade_procedure.md [hunk 17]
```diff
@@ -67,42 +70,143 @@ An upgrade transaction includes a [Plan](https://github.com/cosmos/cosmos-sdk/bl
...
+
+:::
 
 ## Submitting the upgrade on-chain
 
 The `MsgSoftwareUpgrade` can be submitted using the following command:
 
 ```bash
-poktrolld tx authz exec PATH_TO_TRANSACTION_JSON --from pnf
+poktrolld tx authz exec $PATH_TO_UPGRADE_TRANSACTION_JSON --from=pnf
```
[the remark this answers is not in the document]
Comment (okdas) [reply docusaurus/docs/protocol/upgrades/upgrade_procedure.md:122]: This command needs to be adjusted for different networks, I'd prefer the PNF/TestNet Devs to write the full CLI command.

**Comment by okdas** [comment]
@Olshansk I noticed you started reviewing this. I addressed @bryanchriswhite review and pushed more changes to the branch. Just FYI.

**Review by Olshansk — APPROVED** [review]
I pushed a bunch of changes. A couple lingering comments but g2g afterwards!

File: makefiles/localnet.mk [hunk 18]
```diff
@@ -29,3 +29,11 @@ localnet_regenesis: check_yq warn_message_acc_initialize_pubkeys ## Regenerate t
...
 cosmovisor_start_node: ## Starts the node using cosmovisor that waits for an upgrade plan
 	bash tools/scripts/upgrades/cosmovisor-start-node.sh
+
+.PHONY: localnet_cancel_upgrade
+localnet_cancel_upgrade: ## Cancels the planed upgrade on local node
+	poktrolld tx authz exec tools/scripts/upgrades/authz_cancel_upgrade_tx.json --gas=auto --from=pnf
+
+.PHONY: localnet_show_upgrade_plan
+localnet_show_upgrade_plan: ## Shows the upgrade plan on local node
+	poktrolld query upgrade plan
```
[the remark this answers is not in the document]
Comment (okdas) [reply makefiles/localnet.mk:39]: No need. The upgrade happens outside `tilt`/`k8s` deployment - so these commands work as is.

File: makefiles/localnet.mk [hunk 19]
```diff
@@ -29,3 +29,11 @@ localnet_regenesis: check_yq warn_message_acc_initialize_pubkeys ## Regenerate t
 .PHONY: cosmovisor_start_node
 cosmovisor_start_node: ## Starts the node using cosmovisor that waits for an upgrade plan
 	bash tools/scripts/upgrades/cosmovisor-start-node.sh
+
+.PHONY: localnet_cancel_upgrade
+localnet_cancel_upgrade: ## Cancels the planed upgrade on local node
+	poktrolld tx authz exec tools/scripts/upgrades/authz_cancel_upgrade_tx.json --gas=auto --from=pnf
```
[the remark this answers is not in the document]
Comment (okdas) [reply makefiles/localnet.mk:35]: No to both. 
This make target is not executed against a validator running in tilt, but as a standalone validator process running locally, outside k8s deployment.

We specifically instruct turn off LocalNet before testing the upgrade.
